src/client/lazy-app/Modal/index.tsx

```diff
@@ -0,0 +1,136 @@
+import { cleanSet } from '../util/clean-modify';
+import { animateTo } from '../util';
+
+interface Props {
+  icon: VNode;
+  title: string;
+  content: VNode;
+}
+
+interface State {
+  shown: boolean;
+}
+
+export default class Modal extends Component<Props, State> {
+  private dialogElement?: HTMLDialogElement;
+
+  componentDidMount() {
+    if (!this.dialogElement) throw new Error('Modal missing');
+
+    // Set inert by default
+    this.dialogElement.inert = true;
+
+    // Prevent events from leaking through the dialog
+    this.dialogElement.onclick = (event) => {
+      event.preventDefault();
+      event.stopImmediatePropagation();
+    };
+  }
+
+  private _closeOnTransitionEnd = () => {
+    // If modal does not exist
+    if (!this.dialogElement) return;
+
+    this.dialogElement.close();
+    this.dialogElement.inert = true;
+    this.setState({ shown: false });
+  };
+
+  showModal() {
+    if (!this.dialogElement || this.dialogElement.open)
+      throw Error('Modal missing / already shown');
+
+    this.dialogElement.inert = false;
+    this.dialogElement.showModal();
+    // animate modal opening
+    animateTo(
+      this.dialogElement,
+      [
+        { opacity: 0, transform: 'translateY(50px)' },
+        { opacity: 1, transform: 'translateY(0px)' },
+      ],
+      { duration: 250, easing: 'ease' },
+    );
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }, { opacity: 1 }], {
+        duration: 250,
+        easing: 'linear',
+        pseudoElement: '::backdrop',
+      });
+    } catch (e) {}
+    this.setState({ shown: true });
+  }
+
+  private _hideModal() {
+    if (!this.dialogElement || !this.dialogElement.open)
+      throw Error('Modal missing / hidden');
+
+    // animate modal closing
+    const anim = animateTo(
+      this.dialogElement,
+      { opacity: 0, transform: 'translateY(50px)' },
+      { duration: 250, easing: 'ease' },
+    );
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }], {
+        duration: 250,
+        easing: 'linear',
+        pseudoElement: '::backdrop',
+      });
+    } catch (e) {}
+    anim.onfinish = this._closeOnTransitionEnd;
+  }
+
+  private _onKeyDown(e: KeyboardEvent) {
+    // Default behaviour of <dialog> closes it instantly when you press Esc
+    // So we hijack it to smoothly hide the modal
+    if (e.key === 'Escape') {
+      this._hideModal();
+      e.preventDefault();
+      e.stopImmediatePropagation();
+    }
+  }
+
+  render({ title, icon, content }: Props, { shown }: State) {
+    return (
+      <dialog
+        ref={linkRef(this, 'dialogElement')}
+        onKeyDown={(e) => this._onKeyDown(e)}
+      >
+        {shown && (
+          <Fragment>
+            <header class={style.header}>
```

Comment:
If you're going to use `<header>` and `<footer>`, then wrap both in an `<article>`. This semantically links the header and footer.